feat(conanfile): add system_requirements method

ledocc opened the pull request.

ledocc committed:
feat(conanfile): add system_requirements method

ericLemanissier commented:
retriggering CI following testing/20.0.1 branch update

ericLemanissier commented on conanfile.py:
```diff
@@ -268,6 +263,22 @@ def requirements(self):
+            if tools.os_info.with_apt:
+                pack_names = [
+                    "libwayland-dev",
+                    "libwayland-egl-backend-dev",
```
this one does not seem to exist on all platforms: https://github.com/bincrafters/conan-mesa/runs/559437413?check_suite_focus=true#step:5:840

ledocc committed:
fix(conanfile): change class name from LibnameConan to MesaConan

ledocc committed:
fix(conanfile): install wayland-protocols package on ubuntu 16.04 an above, and on Debian 9 and above

ericLemanissier committed:
fix test

ledocc commented:
[gcc 4.9 errors](https://github.com/bincrafters/conan-mesa/pull/2/checks?check_run_id=560976191#step:5:3012) come from [there](https://bugs.freedesktop.org/show_bug.cgi?id=73778), old ubuntu version, old bug.
[other](https://github.com/bincrafters/conan-mesa/pull/2/checks?check_run_id=560976299#step:5:4256) [ones](https://github.com/bincrafters/conan-mesa/pull/2/checks?check_run_id=560976344#step:5:4194) is timeout.

ericLemanissier committed:
enable wayland platform

ericLemanissier committed:
fix distribution names

ledocc committed:
enable wayland build on ubuntu >= 18.04 and debian >= 9

ericLemanissier commented:
Thanks!
One last thing bothers me though: binaries created with this recipe can be different depending on the distro version, and they will still have the same package hash(because same settings and options). An improvement could be to add a with_wayland option with values true,false and auto(the default value). In configure, if with_wayland is auto, we set it to true or false depending on the distro.

ledocc committed:
feat(conanfile): add option with_wayland

Binaries created with this recipe can be different depending on the
distro version, and they will still have the same package hash(because
same settings and options). An improvement could be to add a with_wayland
option with values true,false and auto(the default value). In configure,
if with_wayland is auto, we set it to true or false depending on the distro.

Croydon commented:
Could we instead check if `mako` is installed and raise an exception if not? This would be something we can take over when migrating this to CCI.

ericLemanissier commented:
Well, this is more or less the current situation, the build fails with error `meson.build:805:2: ERROR: Problem encountered: Python (3.x) mako module >= 0.8.0 required to build mesa`, so the message is pretty clear. The only gain we could have is to error earlier (before the sources are downloaded and the build started)

By the way @ledocc, is it actually a good idea to install it in system_requirements ? It looks like mako is needed only to build mesa, not to consume it, right ?

ledocc committed:
fix(conanfile): move mako python install from system_requirements to build_requirements

ledocc committed:
fix(conanfile): search python executable path to install mako package correctly

if conan is run but not installed in virtual environment, sys.executable is
the path of python outside of virtual environment. We want to install mako
inside virtual environment so we have to look for the right python path

ledocc commented:
> By the way @ledocc, is it actually a good idea to install it in system_requirements ? It looks like mako is needed only to build mesa, not to consume it, right ?

done

ericLemanissier commented:
@uilianries and @Croydon what do you think about this PR ?

uilianries commented:
I agree with @Croydon I don't like to see a pip install in a recipe. If this recipe will be migrated someday, and probably will, we won't be able to install mako. Also, I see that pip execution like a system_requirement, which is not allowed in CCI. We can raise a warning message, but make should be installed by the user.

ledocc commented:
ok, 
Installing a python package could be a problem if many recipe do this without care to each other.
But having many recipe that raise Exception because a python package is require will become really boring. 
Conan already handle this {build,system}_dependencies problem for us. We have to find a way to install python package like other {build,system}_dependencies.
Fortunately, python have the virtualenv feature that will fit our need. We can, for each build that have to install python package, create a virtualenv and build package inside.

This have to be done at conan level, so raise a Exception for now and report to conan this problem/solution

Thought ?

uilianries commented:
> This have to be done at conan level, so raise a Exception for now and report to conan this problem/solution

Not exactly, again, it's a python requirement, not a conan package. And system_requirement are not allowed on Conan Center Index for many reasons, and that's why we follow the same approach here.

This recipe is not the first which requires some exotic requirement. You can check if it's installed and raise an ConanException.

ledocc committed:
feat(conanfile): raise exception if mako package in not installed

uilianries commented on conanfile.py:
```diff
@@ -270,6 +278,18 @@ def requirements(self):
+            if pack_names:
+                installer = tools.SystemPackageTool()
+                for item in pack_names:
+                    installer.install(item)
```
Do not install, but only check if it's installed: https://github.com/conan-io/conan-center-index/blob/a504b23f1bd435639e92bd063253ed922577395b/recipes/backward-cpp/all/conanfile.py#L87

ledocc replied:
done in last commit 1f4a941

ledocc replied:
well, I don't understand the first time, done in  85fc51d.
But this is not an usual way to handle missing system package. Currently, conan install missing system package, this simplify CI of package that depend of this one.
Isn't It ?

ledocc replied:
@uilianries now I no longer install system package, many CI build failed ...

[GCC8](https://github.com/bincrafters/conan-mesa/pull/2/checks?check_run_id=745179591#step:5:735)
[GCC9](https://github.com/bincrafters/conan-mesa/pull/2/checks?check_run_id=745179595#step:5:735)
[CLANG6](https://github.com/bincrafters/conan-mesa/pull/2/checks?check_run_id=745179617#step:5:714)
[CLANG7](https://github.com/bincrafters/conan-mesa/pull/2/checks?check_run_id=745179623#step:5:721)
[CLANG8](https://github.com/bincrafters/conan-mesa/pull/2/checks?check_run_id=745179625#step:5:723)
[CLANG9](https://github.com/bincrafters/conan-mesa/pull/2/checks?check_run_id=745179633#step:5:718)

other one is due to a lack of prebuild "m4" package

uilianries replied:
@ledocc The best approach is providing a proxy package or those requirements, you can use this package as reference https://github.com/conan-io/conan-center-index/blob/master/recipes/opengl/all/conanfile.py

Another alternative is installing the packages in CI env, before building. You need to change the CI file.

ledocc replied:
Proxy package looks fun, I'm on it.

ledocc committed:
raise a ConanInvalidConfiguration if required system package is not installed

ledocc committed:
feat(conan): add libglvnd-dev to system package requirement if enable glvnd option
